internal/core/operations/switch_active_version/switch_active_version_executor.go

```diff
@@ -0,0 +1,233 @@
+	logger.Debug("scheduler update finishes with success")
+	return nil
+}
+
+func (ex *SwitchActiveVersionExecutor) OnError(ctx context.Context, op *operation.Operation, definition operations.Definition, executeErr error) error {
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", definition.Name()),
+		zap.String("operation_id", op.ID),
+	)
+	logger.Info("starting OnError routine")
+
+	err := ex.deleteNewCreatedRooms(ctx, logger, op.SchedulerName)
+	ex.clearNewCreatedRooms(op.SchedulerName)
+	if err != nil {
+		return err
+	}
+
+	logger.Debug("finished OnError routine")
+	return nil
+}
+
+func (ex *SwitchActiveVersionExecutor) Name() string {
+	return OperationName
+}
+
+func (ex *SwitchActiveVersionExecutor) replaceRoom(logger *zap.Logger, wg *sync.WaitGroup, roomsChan chan *game_room.GameRoom, roomManager interfaces.RoomManager, scheduler entities.Scheduler) error {
+	defer wg.Done()
+
+	// we're going to use a separated context for each replaceRoom since we
+	// don't want to cancel the replace in the middle (like creating a room and
+	// then left the old one (without deleting it).
+	ctx := context.Background()
+
+	for {
+		room, ok := <-roomsChan
+		if !ok {
+			return nil
+		}
+
+		gameRoom, _, err := roomManager.CreateRoom(ctx, scheduler)
+		if err != nil {
+			ex.roomsBeingReplaced.Delete(room.ID)
+			return err
+		}
+
+		err = roomManager.DeleteRoom(ctx, room)
+		if err != nil {
+			logger.Warn("failed to delete room", zap.Error(err))
+			ex.roomsBeingReplaced.Delete(room.ID)
+			return err
+		}
+
+		logger.Sugar().Debugf("replaced room \"%s\" with \"%s\"", room.ID, gameRoom.ID)
+		ex.roomsBeingReplaced.Delete(room.ID)
+
+		ex.locker.Lock()
+		ex.appendToNewCreatedRooms(scheduler.Name, gameRoom)
+		ex.locker.Unlock()
+	}
+}
+
+func (ex *SwitchActiveVersionExecutor) deleteNewCreatedRooms(ctx context.Context, logger *zap.Logger, schedulerName string) error {
+	logger.Debug("deleting created rooms since switching active version had error - start")
+	for _, room := range ex.newCreatedRooms[schedulerName] {
+		err := ex.roomManager.DeleteRoom(ctx, room)
+		if err != nil {
+			logger.Error("failed to deleted recent created room", zap.Error(err))
+			return err
+		}
+		logger.Sugar().Debugf("deleted room \"%s\" successfully", room.ID)
+	}
+	logger.Debug("deleting created rooms since switching active version had error - end successfully")
+	return nil
+}
+
+func (ex *SwitchActiveVersionExecutor) startReplaceRoomsLoop(ctx context.Context, logger *zap.Logger, maxSurgeNum int, scheduler entities.Scheduler) error {
+	logger.Debug("replacing rooms loop - start")
+	roomsChan := make(chan *game_room.GameRoom)
+	var maxSurgeWg sync.WaitGroup
+	errs, ctx := errgroup.WithContext(ctx)
+
+	maxSurgeWg.Add(maxSurgeNum)
+	for i := 0; i < maxSurgeNum; i++ {
+		errs.Go(func() error {
+			err := ex.replaceRoom(logger, &maxSurgeWg, roomsChan, ex.roomManager, scheduler)
```

Comment:
Can't we just return ex.replaceRoom(logger, &maxSurgeWg, roomsChan, ex.roomManager, scheduler) since we are doing nothing here?

Reply:
You're right!